libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt

```diff
@@ -1,20 +1,98 @@
 import com.mapbox.navigation.base.route.Router
+import com.mapbox.navigation.base.route.model.Route
 import com.mapbox.navigation.base.route.model.RouteOptionsNavigation
+import com.mapbox.navigation.trip.service.NetworkStatusService
+import com.mapbox.navigation.utils.thread.ThreadController
+import com.mapbox.navigation.utils.thread.monitorChannelWithException
+import java.util.concurrent.atomic.AtomicReference
 
 @MapboxNavigationModule(MapboxNavigationModuleType.HybridRouter, skipConfiguration = true)
 class MapboxHybridRouter(
     private val onboardRouter: Router,
-    private val offboardRouter: Router
+    private val offboardRouter: Router,
+    private val context: Context
 ) : Router {
 
+    private interface RouterDispatchInterface {
+        fun execute(routeOptions: RouteOptionsNavigation, clientCallback: Router.Callback)
+    }
+
+    private class OnBoardRouterHandler(private val offRoute: Router, private val onboaredRouter: Router) : RouterDispatchInterface, Router.Callback {
+        private var offrouterCalled = false
+        private var args: RouteOptionsNavigation? = null
+        private var callback: Router.Callback? = null
+
+        override fun onResponse(routes: List<Route>) {
+            callback?.onResponse(routes)
+        }
+
+        override fun onFailure(throwable: Throwable) {
+            when (offrouterCalled) {
+                true -> {
+                    offrouterCalled = false
+                    callback?.onFailure(throwable)
+                }
+                false -> {
+                    offrouterCalled = true
+                    args?.let { options ->
+                        offRoute.getRoute(options, this)
+                    }
+                }
+            }
+        }
+
+        override fun execute(routeOptions: RouteOptionsNavigation, clientCallback: Router.Callback) {
+            offrouterCalled = false
+            args = routeOptions
+            callback = clientCallback
+            onboaredRouter.getRoute(routeOptions, this)
+        }
+    }
+
+    private class OffBoardRouterHandler(private val offboaredRouter: Router) : RouterDispatchInterface, Router.Callback {
+        private var callback: Router.Callback? = null
+        override fun onResponse(routes: List<Route>) {
+            callback?.onResponse(routes)
+        }
+
+        override fun onFailure(throwable: Throwable) {
+            callback?.onFailure(throwable)
+        }
+
+        override fun execute(routeOptions: RouteOptionsNavigation, clientCallback: Router.Callback) {
+            callback = clientCallback
+            offboaredRouter.getRoute(routeOptions, this)
+        }
+    }
+
+    private val navigationMonitorStateException = NetworkStatusService(context)
+    private val jobControl = ThreadController.getIOScopeAndRootJob()
+    private var routeDispatchHandler: AtomicReference<RouterDispatchInterface> = AtomicReference(OnBoardRouterHandler(offboardRouter, onboardRouter))
+
+    init {
```

Comment:
I think it would be better for navigation in this class to move init block and `val`'s to the top of this class body, wdyt?

Reply:
👍